### PR TITLE
[Stylelint] Add invalid properties rule

### PR DESCRIPTION
**Pull request by BSFishy**
### Description

Create a stylelint rule that denies specified properties. We can use this in the future to catch typographic overrides, such as `/font-|line-/` to ensure that OSD and plugins don't override typography. Currently, auto fix simply removes the property.

Right now, I have it set to only find `font-family` properties, with an approved list. A follow up issue will be created to remediate these instances so no font overrides exist in OSD. This is temporary, so we can have the rule and enable it without having failing lints.

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts [hunk 1]
```diff
@@ -27,3 +27,13 @@ export const getSelectorRule = (rules: any, rule: any) => {
 
   return undefined;
 };
+
+export const getPropertyRule = (arr: Record<string, { approved?: string[] }>, decl: any) => {
```
Comment (kavilla) [inline packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts:31]: nit: `arr` is a little bit of misname as it it's a record

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export const getRulesFromConfig = (configPath: string) => {
   return JSON.parse(readFileSync(filePath, 'utf-8'));
 };
 
-export const getSelectorRule = (rules: any, rule: any) => {
+export const getSelectorRule = (rules: Record<string, { approved?: string[] }>, rule: any) => {
```
Comment (kavilla) [inline packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts:21]: nit: might be a little bit over-engineering but we to define a `type` for the two rules config types we have and import and export them within the package?

for example we have a rules config that is files based so in a `index.d.ts` somewhere in the stylelint plugin package:

```typescript
export type FileBasedConfig = Record<string, { approved?: string[] }>;
export type ValueBasedConfig = Record<
  string,
  Record<string, Array<{ approved?: string; rejected?: string[] }>>
>;
```

File: packages/osd-stylelint-config/config/properties.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (kavilla) [inline packages/osd-stylelint-config/config/properties.json:1]: nit: `restricted_properties.json`?

File: packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import {
+  getNotCompliantMessage,
+  getPropertyRule,
+  getRulesFromConfig,
+  isValidOptions,
+} from '../../utils';
+
+const { ruleMessages, report } = stylelint.utils;
+
+const ruleName = 'invalid_properties';
```
Comment (kavilla) [inline packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts:17]: nit: maybe `no_restricted_properties`? Because to me invalid property communicates that this property doesn't exist. But that's just my opinion.

File: packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      };
+
+      if (approvedFiles) {
+        shouldReport = !approvedFiles.some((inspectedFile) => {
+          return file.includes(inspectedFile);
+        });
+      }
+
+      if (shouldReport && isAutoFixing) {
+        decl.remove();
```
Comment (kavilla) [inline packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts:62]: nice! I like it as it removes just the specific property instead of removing the entire style.

I wonder if we backlog an improvement to think about checking the styleblock after this auto fix and deleting the entire style node if empty.

For example,

```scss
.myTitle {
  font-family: 'Courier New'
}
```

will turn into

```scss
.myTitle {
}
```

Which then gets caught by: `block-no-empty` stylelint rule. 

But can be a follow-up,  if we think is needed. The behavior you have of just deleting the property is the correct method here. And I'm fine leaving it as is for now.

**Review by kavilla — APPROVED** [review]
A couple nits to reflect upon but works nicely! Thanks!

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts (hunk 1 above)
Comment (BSFishy) [reply packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts]: Thanks, this was from an old implementation. Updated

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts [hunk 6]
```diff
@@ -27,3 +27,13 @@ export const getSelectorRule = (rules: any, rule: any) => {
 
   return undefined;
 };
+
+export const getPropertyRule = (arr: Record<string, { approved?: string[] }>, decl: any) => {
+  for (const key of Object.keys(arr)) {
+    if (matches(key, decl.prop)) {
```
[the remark this answers is not in the document]
Comment (BSFishy) [reply packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts]: Created generic function

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts (hunk 2 above)
Comment (BSFishy) [reply packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts]: Yep, adding a type for `ValueBasedConfig` also exposed a few cases when things could have been null but weren't tested against

File: packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts (hunk 4 above)
Comment (BSFishy) [reply packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts]: Yeah, I was having trouble coming up with a good name, thanks for the suggestion!

File: packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts [hunk 7]
```diff
@@ -13,15 +13,21 @@ import path from 'path';
 import { readFileSync } from 'fs';
 import { matches } from './matches';
 
+export type FileBasedConfig = Record<string, { approved?: string[] }>;
```
Comment (kavilla) [inline packages/osd-stylelint-plugin-stylelint/src/utils/get_rules_from_config.ts:16]: Nice! Probably can add this to some top level documentation about the two types of supported configs

**Review by kavilla — DISMISSED** [review]
Works awesomely! Thanks!

File: CHANGELOG.md [hunk 8]
```diff
@@ -37,6 +37,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Add an achievement badger to the PR ([#3721](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/3721))
 - Upgrade the backport workflow ([#4343](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4343))
 - [Lint] Add custom stylelint rules and config to prevent unintended style overrides ([#4290](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4290))
+- [Lint] Add stylelint rule to decline configured properties ([#4374](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4374))
```
Comment (joshuarrrr) [inline CHANGELOG.md:40]: Can we rephrase this to be clearer? It seems like the change is that we can now declare CSS properties that can only be specified by allowlisted files. Something along those lines would be better for someone scanning the release notes without context.

**Review by joshuarrrr — DISMISSED** [review]
@BSFishy Is there an issue associated with this? The changelog entry and PR title are very generic, but it sounds like it's partially for https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4314?

File: packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts (hunk 5 above)
Comment (BSFishy) [reply packages/osd-stylelint-plugin-stylelint/src/rules/invalid_properties/index.ts]: I _feel_ like this could be easy, i.e. get parent class, check how many styles after removing, if zero remove class. Need to look into it though.

**Comment by BSFishy** [comment]
> @BSFishy Is there an issue associated with this? The changelog entry and PR title are very generic, but it sounds like it's partially for #4314?

Yes, this is part 1 of #4314. Adding a generic rule where we can deny certain properties, i.e. font related properties. This will help with long-term auditing and preventing new code from adding these properties. We can add another rule down the line that is more advanced and provides more intelligent autofixing/detection.

**Review by joshuarrrr — APPROVED** [review]

**Review by kavilla — APPROVED** [review]
